components/cdc/src/delegate.rs

```diff
@@ -362,12 +362,13 @@ impl Delegate {
             self.region_id,
             change_data_event,
         );
-        for i in 0..downstreams.len() - 1 {
-            if normal_only && downstreams[i].state.load() != DownstreamState::Normal {
+        for curr_downstream in downstreams.iter().take(downstreams.len() - 1) {
+            if normal_only && curr_downstream.state.load() != DownstreamState::Normal {
                 continue;
             }
-            downstreams[i].sink_event(change_data_event.clone());
+            curr_downstream.sink_event(change_data_event.clone());
         }
+        // TODO: also check here ? normal_only and state
```

Comment:
Confirmed that this is a bug, thank you! You can remove this line and change the loop above to include the last item.

Reply:
There's an optimization here where the last `downstream` doesn't receive a copied change_data_event.
This kind of optimization are also seen in other places where the iteration goes to `len - 1` in a loop and the last item is treated outside of the loop.

Should I still move the last downstream in the loop or copy the `if` out for the last downstream ?

Also, what do you think of an issue to review all such iterations to `len - 1` to review any missing checks on the last element ?

Reply:
For DRY, I think the cost of cloning a small object doesn't worth repeating the code. I think it's an anti-pattern.

> Also, what do you think of an issue to review all such iterations to len - 1 to review any missing checks on the last element?

Yes, it would be great if we can review all of them.

Reply:
Fixed and created #8905

Reply:
The save-1-clone (named by me) pattern is so common. How about introducing an iterator adapter for `ExactSizeIterator`, which takes `<T: 'static + Clone, F: FnOnce(T, Iter::Item) -> R, R>(v: T, f: F)` as its parameters?

Reply:
> The save-1-clone (named by me) pattern is so common. How about introducing an iterator adapter for `ExactSizeIterator`, which takes `<T: 'static + Clone, F: Fn(T, Iter::Item) -> R, R>(v: T, f: F)` as its parameters?

Interesting idea, but I think it's out of the scope of this PR.